File: android/Reguerta/app/src/main/java/com/reguerta/user/data/news/ChainedNewsRepository.kt
package com.reguerta.user.data.news

import com.reguerta.user.domain.access.Member
import com.reguerta.user.domain.news.NewsArticle
import com.reguerta.user.domain.news.NewsRepository

class ChainedNewsRepository(
    private val primary: NewsRepository,
    private val fallback: NewsRepository,
) : NewsRepository {
    override suspend fun getNewsFor(member: Member): List<NewsArticle> {
        val primaryNews = primary.getNewsFor(member)
        return if (primaryNews.isNotEmpty()) primaryNews else fallback.getNewsFor(member)
    }

    override suspend fun upsertNews(article: NewsArticle): NewsArticle {
        val fallbackUpdated = fallback.upsertNews(article)
        val primaryUpdated = runCatching { primary.upsertNews(article) }.getOrNull()
        return primaryUpdated ?: fallbackUpdated
    }

    override suspend fun deleteNews(newsId: String): Boolean {
        val fallbackDeleted = fallback.deleteNews(newsId)
        val primaryDeleted = runCatching { primary.deleteNews(newsId) }.getOrDefault(false)
        return primaryDeleted || fallbackDeleted
    }
}

File: ios/Reguerta/Reguerta/Data/News/ChainedNewsRepository.swift
import Foundation

actor ChainedNewsRepository<Primary: NewsRepository, Fallback: NewsRepository>: NewsRepository {
    private let primary: Primary
    private let fallback: Fallback

    init(primary: Primary, fallback: Fallback) {
        self.primary = primary
        self.fallback = fallback
    }

    func news(visibleTo member: Member) async throws -> [NewsArticle] {
        let primaryNews = try await primary.news(visibleTo: member)
        if !primaryNews.isEmpty {
            return primaryNews
        }
        return try await fallback.news(visibleTo: member)
    }

    func allNews() async throws -> [NewsArticle] {
        let primaryNews = try await primary.allNews()
        if !primaryNews.isEmpty {
            return primaryNews
        }
        return try await fallback.allNews()
    }

    func upsert(article: NewsArticle) async throws -> NewsArticle {
        _ = try await fallback.upsert(article: article)
        return try await primary.upsert(article: article)
    }

    func delete(newsId: String) async throws -> Bool {
        let fallbackDeleted = try await fallback.delete(newsId: newsId)
        let primaryDeleted = try await primary.delete(newsId: newsId)
        return primaryDeleted || fallbackDeleted
    }
}
